File: 2020/10.py

import math, copy, re, hashlib
import itertools as it
# import lib for year 2020
from lib import check_data, parse_row, has_all_fields

def rl(arr):
	return range(len(arr))

def part_1(data):
	ones, twos, threes = 0, 0, 0
	data.sort()
	data = [0] + data[::] + [data[-1] + 3]
	i = 1
	while i < len(data):
		if data[i] - data[i - 1] == 1:
			ones += 1
		if data[i] - data[i - 1] == 2:
			twos += 1	
		if data[i] - data[i - 1] == 3:
			threes += 1
		i += 1 
	print(ones, twos, threes)
	print(ones * threes)
	print('END OF PART1')
	return

def get_sums(n, nums):
	if n < 0:
		return 0 
	if n == 1 or n == 0 or len(nums) == 0:
		return 1
	if n == nums[0]:
		return 0
	s1 = get_sums(n - nums[0], nums)
	s2 = get_sums(n, nums[1:])
	return s1 + s2

def part_2(data):
	data.sort()
	data = [0] + data[::] + [data[-1] + 3]
	arr = [1, 2, 3]
	i = 1
	res = 1
	while i < len(data):
		cnt = 0
		while data[i] - data[i - 1] == 1:
			cnt += 1
			i += 1
		if cnt >= 1:
			res *= get_sums(cnt + 1, arr)
		i += 1
	print(res)
	print('END OF PART2')
	return 


if __name__ == '__main__':
	with open('10_input') as f:
		data = f.read()
		# data = data.split('\n')
		data = list(map(int, data.split()))


	part_1(copy.deepcopy(data))
	part_2(copy.deepcopy(data))
	

File: 2022/10.js

const fs = require('fs');
const { exec } = require("child_process");
const { D4, D6, D8, MOD } = require("../lib");
const { cache } = require("../lib");
const { bin, float, hex, int, num, oct } = require("../lib");
const { range, drange, trange, iter, diter, titer } = require("../lib");
const { copy,	entries, in_, inBB, keys, sort, values } = require("../lib");
const { Counter, DD, empty, FastQueue, PriorityQueue, Queue, set, Stack } = require("../lib");
const { ord, chr, count, debug, disp, crt, gcd, lcm, modPow, mod, prod, prodBig, randint, sum, sumBig, transpose } = require("../lib");
const { digits, ints,	floats,	singles,	words,	lines,	table,	groups,	getGroups,	groupsWith,	parse,	parseLine } = require ('../lib');
const { min, max, random, abs, ceil, floor, log, log10, log2, round, sign, sin, cos, tan, asin, acos, atan, atan2, sqrt, PI } = Math;
const { isSuperset, or, and, xor, sub } = set;
const { getExecStr } = require("../lib/post");
const { combinations, combinations_with_replacement, next_permutation, product } = require("../lib/itertools");
// const { parsePacket, ops, HEX, ALU, VARS, resetVars } = require('./lib_2021'); // specific to AOC 2021

function part1(data) {


	let res = 0;
	let cycle = 0;
	let finish = 1;
	let x = 1;
	let next = 20;
	data = lines(data);
	let di = 0;
	let prev = null;
	while (cycle < 220) {
		cycle += 1;
		if (cycle === finish) {
			if (data[di] === 'noop') {
				finish = cycle + 1;
				x += prev !== null ? prev : 0;
				prev = null;
			} else {
				const val = int(data[di].split(' ')[1]);
				x += prev !== null ? prev : 0;
				prev = val;
				finish = cycle + 2;
			}
			++di;
		}
		if (cycle === next) {
			res += next * x;
			next += 40;
		}
	}

	debug(res);
	exec(`echo ${res} | xclip -sel clip -rmlastnl`);
	console.log("END OF PART1");
	return;
}

function part2(data) {

	let cycle = 0;
	let finish = 1;
	let x = 1;
	let next = 40;
	data = lines(data);
	let di = 0;
	let prev = null;
	const img = [];
	const row = [];
	while (cycle < 240) {
		cycle += 1;
		if (x - 1 <= (cycle - 1) % 40 && (cycle - 1) % 40 <= x + 1) {
			row.push('##');
		} else {
			row.push('..');
		}
		if (cycle === finish - 1 && prev !== null) {
			x += prev;
		}
		if (cycle === finish) {
			if (data[di] === 'noop') {
				finish += 1;
				prev = null;
			} else {
				prev = int(data[di].split(' ')[1]);
				finish += 2;
			}
			++di;
		}
		
		if (cycle === next) {
			next += 40;
			img.push([...row].join(''));
			row.length = 0;
		}
	}
	disp(img); // RBPARAGF

	// exec(`echo ${res} | xclip -sel clip -rmlastnl`);
	console.log("END OF PART2");
	return;
}

function main() {
	let data = fs.readFileSync("10_input").toString("utf-8");

	part1(data);
	part2(data);
	process.exit(0);
}

main();
